Andystaples/add functions support

andystaples opened the pull request:
DRAFT PR - for collaboration
Requires at minimum the changes here 
https://github.com/Azure/azure-functions-durable-extension/pull/3260
in durable WebJobs extension allowing gRPC protocol for Python

andystaples committed:
Storing changes commit

andystaples committed:
Working orchestrators + activities

andystaples commented on durabletask-azurefunctions/durabletask/azurefunctions/decorators/durable_app.py:
```diff
@@ -0,0 +1,266 @@
+                if response is None:
+                    raise Exception("Orchestrator execution did not produce a response.")
+                # The Python worker returns the input as type "json", so double-encoding is necessary
+                return '"' + base64.b64encode(response.SerializeToString()).decode('utf-8') + '"'
```
Victoria - Currently, the return value from here is passed on to the host as type "json" so the host attempts to Newtonsoft deserialize it back into an object before handing back to the Durable middleware for final decoding. This breaks, unless I double-encode with quotes as above. Is there a way to communicate to the worker that this is a plain string instead?

hallvictoria replied:
This is actually coming from the OrchestrationTriggerConverter done [here](https://github.com/Azure/azure-functions-python-library/blob/8bd30d300529cbda0526aa2d6606d11333c9c3aa/azure/functions/durable_functions.py#L32). The type is hard-coded to json. Changing the type to `string` works in this scenario, but I'm unsure if there are other cases where `json` was intended

```
@classmethod
    def encode(cls, obj: typing.Any, *, expected_type: typing.Optional[type]) -> meta.Datum:
        # Durable function context should be a json
        return meta.Datum(type='json', value=obj)
```

andystaples commented on durabletask-azurefunctions/durabletask/azurefunctions/decorators/durable_app.py:
```diff
@@ -0,0 +1,266 @@
+        # circumventing a limitation of the worker in type-checking rich DF Client objects.
+        # TODO: Once rich-binding type checking is possible, remove the annotation change.
+        user_code = fb._function._func
+        user_code.__annotations__[parameter_name] = str
```
We'd have to figure out how to detect which underlying provider for the durable_client_input binding is being used to know when to simply return the string for the old SDK vs parse it in the new

andystaples commented on durabletask-azurefunctions/pyproject.toml:
```diff
@@ -0,0 +1,43 @@
+
+[project]
+name = "durabletask.azurefunctions"
+version = "0.1.0"
```
Victoria - what versioning strategy would you propose if the first version that goes to PyPi would be used for internal testing only?

andystaples commented on durabletask-azurefunctions/durabletask/azurefunctions/decorators/durable_app.py:
```diff
@@ -0,0 +1,266 @@
+                def stub_complete(stub_response):
+                    nonlocal response
+                    response = stub_response
+                stub.CompleteOrchestratorTask = stub_complete
```
All of this is probably optimizable - do we really need to create a new stub and worker for each call? Can they be saved? Will look into this more at some point

andystaples committed:
Nitpicks and cleanup

andystaples committed:
Merge branch 'main' into andystaples/add-functions-support

andystaples committed:
Save-all nits

andystaples committed:
Add entity support (needs extension change)

andystaples committed:
Refine entity support
- Still needs eventSent and eventRecieved implementations

andystaples committed:
Finish entity support

andystaples committed:
Fixes and improvements
- Add new_uuid method to OrchestrationContext for deterministic replay-safe UUIDs
- Fix entity locking behavior for Functions
- Align _RuntimeOrchestrationContext param names with OrchestrationContext
- Remap __init__.py files for new module
- Update version to 0.0.1dev0
- Add docstrings to missing methods
- Move code for executing orchestrators/entities to DurableFunctionsWorker
- Add function metadata to triggers for detection by extension

andystaples committed:
Bump durabletask version, fix metadata

andystaples committed:
Use Protocol for stubs

andystaples committed:
Merge branch 'main' into andystaples/add-functions-support

andystaples committed:
Update to new workflow pattern

andystaples committed:
Rename stub file

andystaples committed:
Fix import

andystaples committed:
Merge branch 'main' into andystaples/add-functions-support

andystaples committed:
Experimental dependency revision

andystaples committed:
Update to match changes in functions SDK

andystaples committed:
Merge branch 'main' into andystaples/add-functions-support

andystaples committed:
Merge issue fix
